File: main/binary-tree-postorder-traversal/binary-tree-postorder-traversal-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    static vector<int> postorderTraversal(const TreeNode* root);
};

vector<int> Solution::postorderTraversal(const TreeNode* root)
{
    vector<int> ret;
    const TreeNode* last {}; // The most recent node whose value was emitted.
    
    for (stack<const TreeNode*> s; !s.empty() || root; ) {
        // Go left.
        for (; root; root = root->left) s.push(root);
        
        // Go right by one, if we can and haven't already.
        if (const auto right = s.top()->right) {
            if (right != last) {
                root = right;
                continue;
            }
        }
        
        last = s.top();
        s.pop();
        ret.push_back(last->val);
    }
    
    return ret;
}

File: main/binary-tree-postorder-traversal/binary-tree-postorder-traversal-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        final Deque<TreeNode> path = new ArrayDeque<>();
        final Deque<Boolean> vis = new ArrayDeque<>();

        final Consumer<TreeNode> goLeft = node -> {
            for (; node != null; node = node.left) {
                path.push(node);
                vis.push(Boolean.FALSE);
            }
        };

        final List<Integer> ret = new ArrayList<>();

        for (goLeft.accept(root); !path.isEmpty(); ) {
            if (vis.pop()) {
                ret.add(path.pop().val);
            } else {
                vis.push(Boolean.TRUE);
                goLeft.accept(path.element().right);
            }
        }

        return ret;
    }
}
